beluga/include/beluga/sensor/bearing_sensor_model.hpp

```diff
@@ -0,0 +1,205 @@
+   */
+  template <class Generator>
+  [[nodiscard]] state_type make_random_state(Generator& gen) const {
+    const auto [xmin, xmax, ymin, ymax, zmin, zmax] = landmark_map_.map_limits();
+
+    if constexpr (std::is_same_v<state_type, Sophus::SE3d>) {
+      auto x_distribution = std::uniform_real_distribution<double>{xmin, xmax};
+      auto y_distribution = std::uniform_real_distribution<double>{ymin, ymax};
+      auto z_distribution = std::uniform_real_distribution<double>{zmin, zmax};
+      return Sophus::SE3d{
+          Sophus::SO3d::sampleUniform(gen),
+          Eigen::Vector3d{x_distribution(gen), y_distribution(gen), z_distribution(gen)}};
+    } else {
+      auto x_distribution = std::uniform_real_distribution<double>{xmin, xmax};
+      auto y_distribution = std::uniform_real_distribution<double>{ymin, ymax};
+      return Sophus::SE2d{Sophus::SO2d::sampleUniform(gen), Eigen::Vector2d{x_distribution(gen), y_distribution(gen)}};
+    }
+  }
+
+  /// Gets the importance weight for a particle with the provided state.
+  /**
+   * \param state State of the particle to calculate its importance weight.
+   * \return Calculated importance weight.
+   */
+  [[nodiscard]] weight_type importance_weight(const state_type& state) const {
+    Sophus::SE3d robot_in_world_transform;
+
+    if constexpr (std::is_same_v<state_type, Sophus::SE3d>) {
+      // The robot pose state is already given in 3D,
+      robot_in_world_transform = state;
+    } else {
+      // The robot pose state is given in 2D. Notice that in this case
+      // the 2D pose of the robot is assumed to be that of the robot footprint (projection of the robot
+      // on the z=0 plane of the 3D world frame). This is so that we can tie the sensor reference frame
+      // to the world frame where the landmarks are given without additional structural information.
+      robot_in_world_transform = Sophus::SE3d{
+          Sophus::SO3d::rotZ(state.so2().log()),
+          Eigen::Vector3d{state.translation().x(), state.translation().y(), 0.0}};
+    }
+
+    // precalculate the sensor pose in the world frame
+    const auto sensor_in_world_transform = robot_in_world_transform * params_.sensor_in_robot_transform;
+
+    const auto detection_weight = [this, &sensor_in_world_transform](const auto& sample) {
+      // find the landmark the most closely matches the sample bearing vector
+      const auto landmark_opt = landmark_map_.find_closest_bearing_landmark(sample, sensor_in_world_transform);
+
+      // if we did not find a matching landmark, return 0.0
+      if (!landmark_opt) {
+        return 0.0;
+      }
+
+      // recover sample bearing vector
+      const auto [s_nx, s_ny, s_nz, s_category] = sample;
+
+      // recover landmark bearing vector
+      const auto [l_nx, l_ny, l_nz, l_category] = landmark_opt.value();
+
+      // calculate the aperture vector between the sample and the landmark
+      const auto cos_aperture = s_nx * l_nx + s_ny * l_ny + s_nz * l_nz;
```

Comment:
In this case we were safe because atan2 is insensitive to the module (at worst we were getting K_cos_phi and K_sin_phi, because the landmark bearing was indeed normalized), but to make it safer I made sure the vector is normalized (and also replaced the dot/cross calculation with eigen operations).